packages/react-server/core/components/RootElement.js

```diff
@@ -157,26 +166,36 @@ RootElement.installListener = function(element, listen) {
 }
 
 RootElement.scheduleRender = function(element) {
-	var {listen, when} = (element||{}).props||{};
-	if (!(listen||when)) {
+	var {listen, when, componentLoader, childProps} = (element||{}).props||{};
+	if (!(listen||when||componentLoader||childProps)) {
 		return Q(element).then(RootElement.ensureRootElement);
 	}
 
 	// This is what we'll ultimately resolve our return promise with.
 	// It may be changed by the output of `listen` or `when`.
 	var rendered = element;
+	var componentLoaderDeferred = componentLoader ? componentLoader() : null;
 
 	// Install the listener right away to start gathering props.
 	// It may be a gated emitter, but we want to make sure we squeeze
 	// props out of it from the beginning if it's not.
 	// Finally gate on the `when`.
 	return Q(listen && RootElement.installListener(element, listen))
 		.then(el => el && (rendered = el))
-		.then(() => when)
-		.then(childProps => childProps
-			// merge "when" childProps and "listen" childProps preemptively
-			// to prevent cloneElement shallow merge from clobbering "listen" childProps
-			?React.cloneElement(rendered, {childProps: _.assign({}, rendered.props.childProps || {}, childProps)})
-			:rendered
-		)
+		.then(() => Q.allSettled([when, componentLoaderDeferred]))
+		.then(results => {
+			var [whenResult, loadedComponent] = results;
+			if (whenResult.value || loadedComponent.value || childProps) {
+				// merge in child props from listen, when, and childProps
+				const clonedChildProps = _.assign({}, rendered.props.childProps, whenResult.value, childProps);
```

Comment:
Does this need to also include `rendered.state.childProps` now?

Reply:
I tested it out and it seems that at this point the `constructor` for `RootElement` has not been called (which is to be expected) Thus `rendered.state.childProps` will be `undefined`. However even if the constructor is called, both `rendered.state.childProps` and `rendered.props.childProps` will be the same object and so we are fine with just relying on using `rendered.props.childProps` here. 

I can say that I am confident that `rendered`'s `componentDidMount` will not have been called. That is what hooks up the listener to `listen` which might cause changes to `rendered.state.childProps`.